install/kubernetes: remove all BPF specific flags from default helm

aanm opened the pull request:
Having default values for the specific BPF flags of BPF maps sizes
makes the dynamic size ratio flag to not have any effect. Since the
dynamic size ratio provides a better mechanism for the memory management
dedicated for BPF maps, we should remove the default values entirely
from the default installation of helm.

Clusters being upgraded for which they have each individual flag set the
dynamic ratio flag will not have any effect since each individual flag
takes precedence over the dynamic calculation.

Also, change the default value of `bpf-map-dynamic-size-ratio` to `0.0025` so it is similar to the values set by kube-proxy:

| vCPU                        |        |      1 |       4 |       8 |      64 |      96 |
|-----------------------------|--------|-------:|--------:|--------:|--------:|--------:|
| Memory (GiB)                |        |   3.75 |      15 |      30 |     240 |     360 |
| Kube-proxy CT entries       |        | 131072 |  131072 |  262144 | 2097152 | 3145728 |
| cilium_ct_{4,6}_global      | master | 511847 | 2047388 | 4094776 | 2047388 | 4094776 |
|                             | PR     | 131072 |  131072 |  284560 | 2276480 | 4552960 |
| cilium_ct_{4,6}_any         | master | 255924 | 1023694 | 2047388 | 1023694 | 2047388 |
|                             | PR     | 131072 |  131072 |  142280 | 1138240 | 2276480 |
| cilium_nodeport_neigh{4,6}  | master | 511847 | 2047388 | 4094776 | 2047388 | 4094776 |
|                             | PR     | 131072 |  131072 |  284560 | 2276480 | 4552960 |
| cilium_snat_v{4,6}_external | master | 511847 | 2047388 | 4094776 | 2047388 | 4094776 |
|                             | PR     | 131072 |  131072 |  284560 | 2276480 | 4552960 |
| cilium_lb{4,6}_reverse_sk   | master | 262144 |  262144 |  262144 |  262144 |  262144 |
|                             | PR     | 131072 |  131072 |  142280 | 1138240 | 2276480 |
| cilium_policy_XXXX          | master |  15995 |   63980 |   65535 |   65535 |   65535 |
|                             | PR     |  16384 |  16384 |  16384 | 16384 | 16384 |

```release-note
Remove default bpf map size values for new installations and use the dynamic calculation based on system's memory.
```

Fixes #11964

aanm committed:
pkg/option: set bpf map sizing as info messages

The types of messages are useful overall when the user is starting
Cilium for this reason should be marked as Info and not Debug messages.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
install/kubernetes: remove all BPF specific flags from default helm

Having default values for the specific BPF flags of BPF maps sizes
makes the dynamic size ratio flag to not have any effect. Since the
dynamic size ratio provides a better mechanism for the memory management
dedicated for BPF maps, we should remove the default values entirely
from the default installation of helm.

Clusters being upgraded for which they have each individual flag set the
dynamic ratio flag will not have any effect since each individual flag
takes precedence over the dynamic calculation.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on pkg/option/config.go:
```diff
@@ -2649,23 +2649,23 @@ func (c *DaemonConfig) calculateDynamicBPFMapSizes(totalMemory uint64, dynamicSi
 		c.CTMapEntriesGlobalTCP =
 			getEntries(CTMapEntriesGlobalTCPDefault, LimitTableMin, LimitTableMax)
-		log.Debugf("option %s set by dynamic sizing to %v (default %v)",
+		log.Infof("option %s set by dynamic sizing to %v (default %v)",
```
What does the `default (...)` section of these log messages mean if the default is to dynamically size the maps....?

tklauser replied:
Maybe something like `static default size` would be more appropriate now?

joestringer replied:
I would find it confusing if I deploy Cilium without thinking about this option, then read through the logs and see that it configured the setting to some value then told me the default is a different value.

What are we trying to communicate with that "default" value? Is it a suggested value or something else? Maybe this would be more appropriate on the documentation page for maps & sizes?

tklauser replied:
Good point. These "default" values were mainly useful for debugging during development, so I think it would be fine to drop them from the log message now. Also, these values are already documented in the docs, it seems: https://docs.cilium.io/en/latest/architecture/?highlight=scale#bpf-map-limitations

tgraf replied:
We should remove the `(default [...])` piece

joestringer reviewed (changes requested):
Minor nit below.

tklauser reviewed (commented):
Thank you for fixing this @aanm! LGTM over all, but I agree with @joestringer that we should think about the log message wording in the now that they are `info` rather than `debug` messages. I'll also test this more thoroughly in my setup tomorrow (CEST).

aanm committed:
pkg/option: also account for neigh map in dynamic BPF map calculation

As the neighbor map is one of the top 5 BPF maps that are consuming more
memory we should also add it to the dynamic BPF map calculation.

Fixes: b7be1c000760 ("bpf: make neigh{4,6} maps size configurable from agent and helm")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/option: remove policy map from bpf dynamic map size calculation

The default value for this map only takes roughly 1.5MB of memory for
each endpoint so it does not make sense to have it being calculated in
the dynamic BPF map calculation.

Fixes: 0316bc205d21 ("daemon: add option to dynamically size BPF maps based on system memory")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/option: add sock rev NAT map as part of dynamic BPF map calculation

Since the cilium_lb{4,6}_reverse_sk map has a high memory consumption it
should also be included in the dynamic BPF map calculation.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/option: add a 131072 as minimum number of entries for BPF maps

When performing the dynamic calculation for the BPF maps we should
set 131072 as the minimum number of entries for BPF maps to avoid nodes
with low memory from having issues with our BPF conntrack table.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
install/kubernetes: set default value of BPF dynamic memory to 0.0025

To have a similar behavior seen by kube-proxy we should set the default
value of BPF dynamic memory to 0.0025 which gives us the following
results:

+------+--------------+-----------------------+-------------------+
| vCPU | Memory (GiB) | Kube-proxy CT entries | Cilium CT entries |
+------+--------------+-----------------------+-------------------+
|    1 |         3.75 |                131072 |            131072 |
+------+--------------+-----------------------+-------------------+
|    2 |          7.5 |                131072 |            131072 |
+------+--------------+-----------------------+-------------------+
|    4 |           15 |                131072 |            131072 |
+------+--------------+-----------------------+-------------------+
|    8 |           30 |                262144 |            284560 |
+------+--------------+-----------------------+-------------------+
|   16 |           60 |                524288 |            569120 |
+------+--------------+-----------------------+-------------------+
|   32 |          120 |               1048576 |           1138240 |
+------+--------------+-----------------------+-------------------+
|   64 |          240 |               2097152 |           2276480 |
+------+--------------+-----------------------+-------------------+
|   96 |          360 |               3145728 |           4552960 |
+------+--------------+-----------------------+-------------------+

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/option: do not print default values of BPF maps

This can be misleading for users so we should avoid printing those.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on pkg/option/config.go:
```diff
@@ -474,6 +474,10 @@ const (
 
+	// LimitTableAutoMin defines the minimum CT or NAT table limit for dynamic
+	// size ration calculation.
+	LimitTableAutoMin = 1 << 17 // 128Ki entries
```
Please rename this to something that is specific to those tables.

tgraf commented on pkg/option/config_test.go:
```diff
@@ -779,23 +779,23 @@ func TestBPFMapSizeCalculation(t *testing.T) {
-				NATMapSize:        56912,
-				NeighMapSize:      56912,
-				SockRevNatMapSize: 28456,
+				CTMapSizeTCP:      LimitTableAutoMin,
```
I think we should have a const for each table even if it means that some of them have the same value. Anybody changing `LimitTableAutoMin` might not understand what consequences it has.

tklauser reviewed (commented):
cmdref will also need updating, see the Travis CI failure:

```
--- a/Documentation/cmdref/cilium-agent.md
+++ b/Documentation/cmdref/cilium-agent.md
@@ -38,6 +38,7 @@ cilium-agent [flags]
       --bpf-neigh-global-max int                      Maximum number of entries for the global BPF neighbor table (default 524288)
       --bpf-policy-map-max int                        Maximum number of entries in endpoint policy map (per endpoint) (default 16384)
       --bpf-root string                               Path to BPF filesystem
+      --bpf-sock-rev-map-max int                      Maximum number of entries for the SockRevNAT BPF map (default 262144)
       --certificates-directory string                 Root directory to find certificates specified in L7 TLS policy enforcement (default "/var/run/cilium/certs")
       --cgroup-root string                            Path to Cgroup2 filesystem
       --cluster-id int                                Unique identifier of the cluster
HINT: to fix this, run 'make -C Documentation update-cmdref ; git commit Documentation/cmdref --message "docs: Update cmdref"'
Makefile:32: recipe for target 'check' failed
```

tklauser commented on pkg/option/config.go:
```diff
@@ -496,6 +501,10 @@ const (
 
+	// SockRevNatEntriesName configures max entries for BPF sock reverse nat
+	// entries.
+	SockRevNatEntriesName = "bpf-sock-rev-map-max"
```
I think this should also be added to the "BPF flags" section in `HelpFlagSections` in the same file, so it appears with the other map size flags in the help output and cmdref.

nebril reviewed (approved):
Helm bits LGTM

tklauser reviewed (approved):
~I think the cmdref needs an update again after the move of `--bpf-sock-rev-map-max` to the "BPF flags" section. LGTM other than that.~

joestringer commented on pkg/option/config.go:
```diff
@@ -2677,44 +2732,44 @@ func (c *DaemonConfig) calculateDynamicBPFMapSizes(totalMemory uint64, dynamicSi
 	} else {
 		log.Debugf("option %s set by user to %v", NeighMapEntriesGlobalName, c.NeighMapEntriesGlobal)
 	}
-	if !viper.IsSet(PolicyMapEntriesName) {
```
What happened to policymap autosizing configuration..?

aanm replied:
It's in the commit message :-) its size was irrelevant compared to the other ones that's why I've remove it

joestringer commented on Documentation/architecture.rst:
```diff
@@ -271,6 +271,45 @@ connection tracking, NAT, and policy maps are determined at agent startup based
 on the given ratio of the total system memory. For example a given ratio of 0.03
 leads to 3% of the total system memory to be used for these maps.
 
+This flag sets the ratio of total system memory to use for dynamic size of the
```
The documentation table just a few paragraphs above here list a bunch of default limits which also need updating with this PR.

aanm replied:
The defaults didn't change, if they did that would be a bug. If the dynamic sizing is `0.0` the defaults set in the binary are used instead.
